### ISSUE TITLE
semi-supervised hierarchical mode for BERTopic

### ISSUE THREAD

**Issue by fancytony** [MaartenGr/BERTopic#920]
Hey Maarten,

I am writing to ask if such a semi-supervised hierarchical mode is available for BERTopic or not. The problem I am working on is we already have some known categories (like the 20 categories for the newsgroup data) but look for generating sub-categories for each category. An example would be we have labels like Automobiles, Sports .etc, but try to mine the data to obtain 'Sedan, SUV, Coupe...' and 'volleyball, tennis...'. Of course, one way to do this is to generate one model for each category; but that seems awkward and creates maintenance issues (N models instead of 1 model); The ability of semi-supervised hierarchical topic modeling is supported by other algorithms such as Semi-Supervised Hierarchical LDA (see this paper SSHLDA: [A Semi-Supervised Hierarchical Topic Model](https://aclanthology.org/D12-1073.pdf). Is there a way to achieve this using 1 model to generate the whole tree (with the first layer of tree nodes fixed to be the known labels)? I am sure there are a lot of such use scenarios. any feedback/comments would be greatly appreciated.

BTW, thank you so much for developing the BERTopic (and I am surprised to find you have been keeping updating the docs:-).  I find BERTopic is super-valuable and can solve a lot of topic-modeling problems! Great work Maarten!

**Comment by fancytony** [comment]
<img width="372" alt="image" src="https://user-images.githubusercontent.com/25394728/211680442-2920a019-0893-4ab2-8f6c-83d466cf6a58.png">
adding an example from the above paper to illustrate the case. thanks.

**Comment by MaartenGr** [comment]
Nice timing! I had been working on it since a few days ago. 

Although it is not currently implemented, there are a few things that you can do to approach this as much as possible. To start off, as the name implies, there are two methodologies to this, namely **semi-supervised** topic modeling and **hierarchical** topic modeling. Both methodologies are implemented in BERTopic and we can run semi-supervised first and then on top of that hierarchical topic modeling. There are a few things though that we need to take into account which I will go into a bit more detail below. 

## Prepare Data
First, let's prepare our data and embeddings:

```python
from sentence_transformers import SentenceTransformer
from sklearn.datasets import fetch_20newsgroups

# Get labeled data
data = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))
docs = data['data']
y = data['target']
target_names = data["target_names"]

# Prepare embeddings
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(docs, show_progress_bar=True)
```

## Semi-supervised
Semi-supervised topic modeling will only nudge the dimensionality reduction process toward the given labels. For semi-supervised hierarchical topic model, has both advantages and disadvantages. An advantage is that the topic model will most likely find more topics than the labels and these are likely to be sub-topics that you are looking for. A disadvantage is that we want to maximize the separation between labels whilst still keeping some structure within those labels. Only slightly nudging can create overlap between those clusters. 

What we can do to resolve that is by fine-tuning the embeddings based on the classes. It allows us to extend this nudging process whilst, hopefully, still keeping the separation. There are a few ways to approach this, from straightforward/naive to increasingly complex/accurate:

* We can simply average the embeddings of class with that classes averaged embeddings
* We can fine-tune the embeddings with something like [embetter](https://koaning.github.io/embetter/finetuners/)
* We can fine-tune the embedding model itself using contrastive learning in [sentence-transformers](https://github.com/UKPLab/sentence-transformers/) or even [SetFit](https://github.com/huggingface/setfit)
  * This can be left up to the reader as it is often a bit more involved

#### Averaged Embeddings
After creating our embeddings, we can create averaged embeddings for each target. Then, we can average those averaged embeddings with the original embeddings: 

```python
embeddings_averaged = embeddings.copy()
for target in set(y):
    indices = [index for index, t in enumerate(y) if t==target]
    embeddings_averaged[indices] = np.average([embeddings_averaged[indices], np.mean(embeddings_averaged[indices])], weights=[1, 1])
```

We can use these embeddings as the input for our topic model. 

#### Embetter
Without going into too much detail, we can put a small neural network on top of our embeddings and try to predict the labels. The hidden layers can then be used as our updated embeddings and should be fine-tuned towards our labels. You can find more about the process [here](https://koaning.github.io/embetter/finetuners/). 

```python
# from embetter.finetune import ForwardFinetuner 
tuner = ForwardFinetuner(n_epochs=500, learning_rate=0.01, hidden_dim=300)

# Note that we can do this also for the `embeddings_averaged` above!
tuner.fit(embeddings, y)
embetter_embeddings = tuner.transform(embeddings)
```

We can use these embeddings as the input for our topic model. 

#### 2D-space

We can plot these embeddings into 2D-space to view the potential separation:

```python
from umap import UMAP
import seaborn as sns
import matplotlib.pyplot as plt
embedding_sets = [(embeddings, "Embeddings"), (embeddings_averaged, "Averaged Embeddings"), 
                  (embetter_embeddings, "Embetter Embeddings"), (embetter_embeddings_from_averaged, "Embetter Embeddings from Averaged")]
fig, ax = plt.subplots(1,len(embedding_sets), figsize=(60, 15))

# Reduce dimensionality and plot
for index, (embedding_set, title) in enumerate(embedding_sets):
    reduced_embeddings = UMAP(n_neighbors=10, n_components=2, metric="cosine", random_state=42).fit_transform(embedding_set, y=y)
    sns.scatterplot(x=reduced_embeddings[:, 0], y=reduced_embeddings[:, 1], hue=y, ax=ax[index], alpha=0.3, size=1, palette=sns.color_palette("tab20"))
    ax[index].set_title(title, fontdict={"fontsize": 50})
fig.show()
```

![results (1)](https://user-images.githubusercontent.com/25746895/211754453-7302a86e-81bd-4ce6-908b-4c0d9974e33e.png)

You can open up the image and zoom in a bit but it seems that the [embetter](https://github.com/koaning/embetter) does a bit better than the others in keeping the labels within the same cluster. There also seem to be fewer outliers in 2D-space here. This effect is much more pronounced if we would not have set `y=y` in UMAP but since this is what we are using in BERTopic, this seemed the most accurate. 


## Hierarchical Topic Modeling
What we need to take into account when performing hierarchical topic modeling is that we cannot be sure that the topics found above will be merged back into the classes they initially contained. They are not tracked in that way, so we need to find a way to match (merged) topics with the initial labels. To do so, we can try, for each (merged) topic, to find how many documents overlap with the documents for a specific label:

```python
results = pd.DataFrame(columns = ["Parent_ID", "Topics", "Target", "Overlap1", "Overlap2", "Score", "Score1"])

# Add merged topics
for parent_id, topic_set in tqdm(zip(hierarchical_topics.Parent_ID.values, hierarchical_topics.Topics.values)):
    hierarchy_indices = {index for topic in topic_set for index, val in enumerate(topics) if val == topic}
    
    for target in set(y):
        y_indices = set([index for index, t in enumerate(y) if t == target])
        overlap1 = round(len(hierarchy_indices.intersection(y_indices)) / len(y_indices) * 100, 1)
        overlap2 = round(len(hierarchy_indices.intersection(y_indices)) / len(hierarchy_indices) * 100, 1)
        score = round(np.mean((overlap1, overlap2)), 1)
        score1 = round(min(overlap1, overlap2), 1)
        results.loc[len(results), :] = [parent_id, topic_set, target, overlap1, overlap2, score, score1]
        
# Add unmerged topics
for parent_id, topic_set in tqdm(zip(set(topics), [[topic] for topic in set(topics)])):
    hierarchy_indices = {index for topic in topic_set for index, val in enumerate(topics) if val == topic}
    
    for target in set(y):
        y_indices = set([index for index, t in enumerate(y) if t == target])
        overlap1 = round(len(hierarchy_indices.intersection(y_indices)) / len(y_indices) * 100, 1)
        overlap2 = round(len(hierarchy_indices.intersection(y_indices)) / len(hierarchy_indices) * 100, 1)
        score = round(np.mean((overlap1, overlap2)), 1)
        score1 = round(min(overlap1, overlap2), 1)
        results.loc[len(results), :] = [parent_id, topic_set, target, overlap1, overlap2, score, score1]

results = results[results['Score']==results.groupby('Target')['Score'].transform('max')]
```


## Full Example
Now, let's combine it all together to get our semi-supervised hierarchical topic model. First, we need to prepare our embeddings. For the sake of simplicity, I will be using the averaged embeddings:

```python
import numpy as np
from sklearn.datasets import fetch_20newsgroups
from sentence_transformers import SentenceTransformer

# Get labeled data
data = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))
docs = data['data']
y = data['target']
target_names = data["target_names"]

# Prepare embeddings
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(docs, show_progress_bar=True)

# Average embeddings based on the class
embeddings_averaged = embeddings.copy()
for target in set(y):
    indices = [index for index, t in enumerate(y) if t==target]
    embeddings_averaged[indices] = np.average([embeddings_averaged[indices], np.mean(embeddings_averaged[indices])], weights=[1, 1])
```

Then, we can perform our semi-supervised topic model. However, we need to replace HDBSCAN with an algorithm that does not generate any outliers since our labels also do not contain those:

```python
from umap import UMAP
from bertopic import BERTopic
from sklearn.cluster import KMeans
from sklearn.feature_extraction.text import CountVectorizer

# Sub models
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42)
cluster_model = KMeans(n_clusters=150)
vectorizer_model = CountVectorizer(stop_words="english")

# Train our semi-supervised topic model
topic_model = BERTopic(hdbscan_model=cluster_model, umap_model=umap_model, vectorizer_model=vectorizer_model)
topics, probs = topic_model.fit_transform(docs, embeddings_averaged, y=y)
```

We train our hierarchical topic model and we find the (merged) topics that belong to our initial labels:

```python
import pandas as pd
from tqdm import tqdm
results = pd.DataFrame(columns = ["Parent_ID", "Topics", "Target", "Overlap1", "Overlap2", "Score", "Score1"])

# Add merged topics
for parent_id, topic_set in tqdm(zip(hierarchical_topics.Parent_ID.values, hierarchical_topics.Topics.values)):
    hierarchy_indices = {index for topic in topic_set for index, val in enumerate(topics) if val == topic}
    
    for target in set(y):
        y_indices = set([index for index, t in enumerate(y) if t == target])
        overlap1 = round(len(hierarchy_indices.intersection(y_indices)) / len(y_indices) * 100, 1)
        overlap2 = round(len(hierarchy_indices.intersection(y_indices)) / len(hierarchy_indices) * 100, 1)
        score = round(np.mean((overlap1, overlap2)), 1)
        score1 = round(min(overlap1, overlap2), 1)
        results.loc[len(results), :] = [parent_id, topic_set, target, overlap1, overlap2, score, score1]
        
# Add unmerged topics
for parent_id, topic_set in tqdm(zip(set(topics), [[topic] for topic in set(topics)])):
    hierarchy_indices = {index for topic in topic_set for index, val in enumerate(topics) if val == topic}
    
    for target in set(y):
        y_indices = set([index for index, t in enumerate(y) if t == target])
        overlap1 = round(len(hierarchy_indices.intersection(y_indices)) / len(y_indices) * 100, 1)
        overlap2 = round(len(hierarchy_indices.intersection(y_indices)) / len(hierarchy_indices) * 100, 1)
        score = round(np.mean((overlap1, overlap2)), 1)
        score1 = round(min(overlap1, overlap2), 1)
        results.loc[len(results), :] = [parent_id, topic_set, target, overlap1, overlap2, score, score1]
        
results = results[results['Score']==results.groupby('Target')['Score'].transform('max')]
```

<img src="https://user-images.githubusercontent.com/25746895/211759083-1b869eba-0f72-4f33-b859-2afddf784386.png"  height="400" />

Based on the table above, we can see that label 14 which relates to NewsGroups class `'sci.space'` has a few sub-topics, namely 
`[11, 21, 30, 47, 90, 94]` which are the following:

* '11 - space - launch - moon - lunar - nasa'
* '21 - space - nasa - venus - solar - spacecraft'
* '30 - mission - shuttle - orbit - spacecraft - hst'
* '47 - universe - oort - theory - larson - launch'
* '90 - objective - theory - science - universe - larsons'
* '94 - kirlian - photography - science - methodology - leaf'

**Comment by fancytony** [comment]
Thank you so much, Maarten!  Appreciate the quick and detailed response, really helpful! 

BTW, in case someone else also bumped into this thread, I think you missed a line in the code (which will cause error if not added):
`hierarchical_topics = topic_model.hierarchical_topics(docs)`

Thanks again!

**Comment by MaartenGr** [comment]
Thanks, I'll make sure to update the code above!

**Comment by MaartenGr** [comment]
@lintonye If you only know the category names but not the category for each doc, then you could use either [guided BERTopic](https://maartengr.github.io/BERTopic/getting_started/guided/guided.html) or you could assign those categories to each document through cosine similarity between categories and document embeddings.
